### ISSUE TITLE
Python behind the scenes #1: how the CPython VM works

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7427]
Python behind the scenes #1: how the CPython VM works<br>
<br>
https://ift.tt/3jJDvwo<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<h2>Introduction</h2>
<p>Have you ever wondered what <code>python</code> does when you run one of your programs?</p>
<p>This article opens a series which seeks to answer this very question. We'll dive into the internals of CPython, Python's most popular implementation. By doing so we'll understand the language itself at a deeper level. That is the primary goal of this series. If you're familiar with Python and comfortable reading C but have no much experience working with CPython's source code, there is a good chance you'll find this writing interesting.</p>
<h3>What CPython is and why anyone would want to study it</h3>
<p>Let's begin by stating some well-known facts. CPython is a Python interpreter written in C. It's one of the Python implementations, alongside with PyPy, Jython, IronPython and many others. CPython is distinguished in that it is original, most-maintained and the most popular one.</p>
<p>CPython implements Python, but what is Python? One may simply answer &ndash; Python is a programming language. The answer becomes much more nuanced when the same question is put properly: what defines what Python is? Python, unlike languages like C, doesn't have a formal specification. The thing that comes closest to it is the <a href="https://docs.python.org/3.9/reference/index.html">Python Language Reference</a> which starts with the following words:</p>
<blockquote>
<p>While I am trying to be as precise as possible, I chose to use English rather than formal specifications for everything except syntax and lexical analysis. This should make the document more understandable to the average reader, but will leave room for ambiguities. Consequently, if you were coming from Mars and tried to re-implement Python from this document alone, you might have to guess things and in fact you would probably end up implementing quite a different language. On the other hand, if you are using Python and wonder what the precise rules about a particular area of the language are, you should definitely be able to find them here.</p>
</blockquote>
<p>So Python is not defined by its language reference only. It would be also wrong to say that Python is defined by its reference implementation, CPython, since there are some implementation details that are not a part of the language. A garbage collector that relies on a reference counting is one example. Since there is no single source of truth, we may say that Python is defined partly by the Python Language Reference and partly by its main implementation, CPython.</p>
<p>Such a reasoning may seem pedantic, but I think it is crucial to clarify the key role of the subject we're going to study. You might still wonder, though, why we should study it. Besides plain curiosity, I see the following reasons:</p>
<ul>
<li>Having a full picture gives a deeper understanding of the language. It's much more easier to grasp some peculiarity of Python if you're aware of its implementation details.</li>
<li>Implementation details matter in practice. How objects are stored, how the garbage collector works and how multiple threads are coordinated are subjects of high importance when one wants to understand the of applicability of the language and its limitations, estimate performance or detect inefficiencies.</li>
<li>CPython provides Python/C API which allows to extend Python with C and embed Python inside C. To use this API effectively a programmer needs a good understanding of how CPython works.</li>
</ul>
<h3>What it takes to understand how CPython works</h3>
<p>CPython was designed to be easy to maintain. A newcomer can certainly expect to be able to read the source code and understand what it does. However, it may take some time. By writing this series I hope to help you to shorten it.</p>
<h3>How this series is laid out</h3>
<p>I chose to take a top-down approach. In this part we'll explore the core concepts of the CPython virtual machine (VM). Next, we'll see how CPython compiles a program into something that the VM can execute. After that, we'll get familiar with the source code and step through the execution of a program studying main parts of the interpreter on the way. Eventually, we'll be able to pick out different aspects of the language one by one and see how they're implemented. This is by no way a strict plan but my approximate idea.</p>
<p><strong>Note</strong>: In this post I'm referring to CPython 3.9. Some implementation details will certainly change as CPython evolves. I'll try to keep track of important changes and add update notes.</p>
<h2>The big picture</h2>
<p>An execution of a Python program roughly consists of three stages:</p>
<ol>
<li>Initialization</li>
<li>Compilation</li>
<li>Interpretation</li>
</ol>
<p>During initialization stage, CPython initializes data structures required to run Python. It also prepares such things as builtin types, configures and loads builtin modules, sets up the import system and does many other things. This is very important stage that is often overlooked by the CPython's explorers because of its service nature.</p>
<p>Next comes the compilation stage. CPython is an interpreter, not a compiler in a sense that it doesn't produce machine code. Interpreters, however, usually translate source code into some intermediate representation before executing it. So does CPython. This translation phase does the same things a typical compiler does: parses a source code and builds an AST (Abstract Syntax Tree), generates bytecode from an AST and even performs some bytecode optimizations.</p>
<p>Before looking at the next stage, we need to understand what bytecode is. Bytecode is a series of instructions. Each instruction consists of two bytes: one for an opcode and one for an argument. Consider an example:</p>
<p>CPython translates the body of the function <code>g</code> to the following sequence of bytes: <code>[124, 0, 100, 1, 23, 0, 83, 0]</code>. If we run a standard <a href="https://docs.python.org/3/library/dis.html"><code>dis</code></a> module to disassemble it, here's what we'll get:</p>
<div>
<pre>$ python -m dis example1.py
...
<span>2</span>           <span>0</span> LOAD_FAST            <span>0</span> <span>(</span>x<span>)</span>
            <span>2</span> LOAD_CONST           <span>1</span> <span>(</span><span>3</span><span>)</span>
            <span>4</span> BINARY_ADD
            <span>6</span> RETURN_VALUE
</pre>
</div>
<p><code>LOAD_FAST</code> opcode corresponds to a byte <code>124</code> and has an argument <code>0</code>. <code>LOAD_CONST</code> opcode corresponds to a byte <code>100</code> and has an argument <code>1</code>. <code>BINARY_ADD</code> and <code>RETURN_VALUE</code> instructions are always encoded as <code>(23, 0)</code> and <code>(83, 0)</code> respectively since they don't need an argument.</p>
<p>At the heart of CPython is a virtual machine that executes bytecode. By looking at the previous example you might guess how it works. CPython's VM is stack-based. It means that it executes instructions using the stack to store and retrieve data. <code>LOAD_FAST</code> instruction pushes local variable on the stack. <code>LOAD_CONST</code> pushes a constant. <code>BINARY_ADD</code> pops two objects from the stack, adds them up and pushes the result back. Finally, <code>RETURN_VALUE</code> pops whatever is on the stack and returns the result to its caller.</p>
<p>The bytecode execution happens in a giant evaluation loop that runs while there are instructions to execute. It stops to yield a value or if an error occured.</p>
<p>Such a brief overview gives rise to a lot of questions:</p>
<ul>
<li>What do the arguments to the opcodes <code>LOAD_FAST</code> and <code>LOAD_CONST</code> mean? Are they indices? What do they index?</li>
<li>Does the VM places values or references to the objects on the stack?</li>
<li>How does CPython know that <code>x</code> is a local variable?</li>
<li>What if an argument is too big to fit into a single byte?</li>
<li>Is instruction for adding up two numbers the same as for concatenating two strings? If yes, then how does the VM differentiate between these operations?</li>
</ul>
<p>In order to answer these and other intriguing questions we need to look at the core concepts of the CPython VM.</p>
<h3>Code objects, function objects, frames</h3>
<h4>code object</h4>
<p>We saw what bytecode of a simple function looks like. But a typical Python program is more complicated. How does the VM execute a module that contains function definitions and make function calls?</p>
<p>Consider a program:</p>
<div>
<pre><span>def</span> <span>f</span><span>(</span><span>x</span><span>):</span>
    <span>return</span> <span>x</span> <span>+</span> <span>1</span>

<span>print</span><span>(</span><span>f</span><span>(</span><span>1</span><span>))</span>
</pre>
</div>
<p>What does its bytecode look like? To answer this question, let's analyze what the program does. It defines a function <code>f</code>, calls function <code>f</code> with <code>1</code> as an argument and prints the result of the call. Whatever the function <code>f</code> does, it's not a part of the module's bytecode. We can assure ourselves by running a disassembler.</p>
<div>
<pre>$ python -m dis example2.py

<span>1</span>           <span>0</span> LOAD_CONST               <span>0</span> <span>(</span>&lt;code object f at 0x10bffd1e0, file <span>"example.py"</span>, line <span>1</span>&gt;<span>)</span>
            <span>2</span> LOAD_CONST               <span>1</span> <span>(</span><span>'f'</span><span>)</span>
            <span>4</span> MAKE_FUNCTION            <span>0</span>
            <span>6</span> STORE_NAME               <span>0</span> <span>(</span>f<span>)</span>

<span>4</span>           <span>8</span> LOAD_NAME                <span>1</span> <span>(</span>print<span>)</span>
           <span>10</span> LOAD_NAME                <span>0</span> <span>(</span>f<span>)</span>
           <span>12</span> LOAD_CONST               <span>2</span> <span>(</span><span>1</span><span>)</span>
           <span>14</span> CALL_FUNCTION            <span>1</span>
           <span>16</span> CALL_FUNCTION            <span>1</span>
           <span>18</span> POP_TOP
           <span>20</span> LOAD_CONST               <span>3</span> <span>(</span>None<span>)</span>
           <span>22</span> RETURN_VALUE
...
</pre>
</div>
<p>On line one we define function <code>f</code> by making a function from something called code object and binding name <code>f</code> to it. We don't see the bytecode of the function <code>f</code> that returns an incremented argument.</p>
<p>The pieces of code that are executed as a single unit like a module or a function body are called code blocks. CPython stores information about what a code block does in a structure called code object. It contains the bytecode and such things as lists of names of variables used within the block. To run a module or to call a function means to start evaluating a corresponding code object.</p>
<h4>function object</h4>
<p>A function, however, is not merely a code object. It must include additional information such as name, docstring, default arguments and values of variables defined in the enclosing scope. This information, together with a code object, is stored inside a function object. <code>MAKE_FUNCTION</code> instruction is used to create it. The definition of the function object structure in the CPython source code is preceeded by the following comment:</p>
<blockquote>
<p>Function objects and code objects should not be confused with each other:</p>
<p>Function objects are created by the execution of the 'def' statement. They reference a code object in their __code__ attribute, which is a purely syntactic object, i.e. nothing more than a compiled version of some source code lines. There is one code object per source code "fragment", but each code object can be referenced by zero or many function objects depending only on how many times the 'def' statement in the source was executed so far.</p>
</blockquote>
<p>How can it be that several function objects reference a single code object? Here is an example:</p>
<div>
<pre><span>def</span> <span>make_add_x</span><span>(</span><span>x</span><span>):</span>
    <span>def</span> <span>add_x</span><span>(</span><span>y</span><span>):</span>
        <span>return</span> <span>x</span> <span>+</span> <span>y</span>
    <span>return</span> <span>add_x</span>

<span>add_4</span> <span>=</span> <span>make_add_x</span><span>(</span><span>4</span><span>)</span>
<span>add_5</span> <span>=</span> <span>make_add_x</span><span>(</span><span>5</span><span>)</span>
</pre>
</div>
<p>The bytecode of the <code>make_add_x</code> function contains a <code>MAKE_FUNCTION</code> instruction. The functions <code>add_4</code> and <code>add_5</code> are the result of calling this instruction with the same code object as an argument. But there is one argument that differs &ndash; the value of <code>x</code>. Each function gets its own by the mechanism of cell variables that allows us to create closures like <code>add_4</code> and <code>add_5</code>.</p>
<p>I recommend you to take a look at the definitions of the code and function objects before we move to the next concept.</p>
<div>
<pre><span>struct</span> <span>PyCodeObject</span> <span>{</span>
    <span>PyObject_HEAD</span>
    <span>int</span> <span>co_argcount</span><span>;</span>            <span>/* #arguments, except *args */</span>
    <span>int</span> <span>co_posonlyargcount</span><span>;</span>     <span>/* #positional only arguments */</span>
    <span>int</span> <span>co_kwonlyargcount</span><span>;</span>      <span>/* #keyword only arguments */</span>
    <span>int</span> <span>co_nlocals</span><span>;</span>             <span>/* #local variables */</span>
    <span>int</span> <span>co_stacksize</span><span>;</span>           <span>/* #entries needed for evaluation stack */</span>
    <span>int</span> <span>co_flags</span><span>;</span>               <span>/* CO_..., see below */</span>
    <span>int</span> <span>co_firstlineno</span><span>;</span>         <span>/* first source line number */</span>
    <span>PyObject</span> <span>*</span><span>co_code</span><span>;</span>          <span>/* instruction opcodes */</span>
    <span>PyObject</span> <span>*</span><span>co_consts</span><span>;</span>        <span>/* list (constants used) */</span>
    <span>PyObject</span> <span>*</span><span>co_names</span><span>;</span>         <span>/* list of strings (names used) */</span>
    <span>PyObject</span> <span>*</span><span>co_varnames</span><span>;</span>      <span>/* tuple of strings (local variable names) */</span>
    <span>PyObject</span> <span>*</span><span>co_freevars</span><span>;</span>      <span>/* tuple of strings (free variable names) */</span>
    <span>PyObject</span> <span>*</span><span>co_cellvars</span><span>;</span>      <span>/* tuple of strings (cell variable names) */</span>

    <span>Py_ssize_t</span> <span>*</span><span>co_cell2arg</span><span>;</span>    <span>/* Maps cell vars which are arguments. */</span>
    <span>PyObject</span> <span>*</span><span>co_filename</span><span>;</span>      <span>/* unicode (where it was loaded from) */</span>
    <span>PyObject</span> <span>*</span><span>co_name</span><span>;</span>          <span>/* unicode (name, for reference) */</span>
        <span>/* ... more members ... */</span>
<span>};</span>
</pre>
</div>
<div>
<pre><span>typedef</span> <span>struct</span> <span>{</span>
    <span>PyObject_HEAD</span>
    <span>PyObject</span> <span>*</span><span>func_code</span><span>;</span>        <span>/* A code object, the __code__ attribute */</span>
    <span>PyObject</span> <span>*</span><span>func_globals</span><span>;</span>     <span>/* A dictionary (other mappings won't do) */</span>
    <span>PyObject</span> <span>*</span><span>func_defaults</span><span>;</span>    <span>/* NULL or a tuple */</span>
    <span>PyObject</span> <span>*</span><span>func_kwdefaults</span><span>;</span>  <span>/* NULL or a dict */</span>
    <span>PyObject</span> <span>*</span><span>func_closure</span><span>;</span>     <span>/* NULL or a tuple of cell objects */</span>
    <span>PyObject</span> <span>*</span><span>func_doc</span><span>;</span>         <span>/* The __doc__ attribute, can be anything */</span>
    <span>PyObject</span> <span>*</span><span>func_name</span><span>;</span>        <span>/* The __name__ attribute, a string object */</span>
    <span>PyObject</span> <span>*</span><span>func_dict</span><span>;</span>        <span>/* The __dict__ attribute, a dict or NULL */</span>
    <span>PyObject</span> <span>*</span><span>func_weakreflist</span><span>;</span> <span>/* List of weak references */</span>
    <span>PyObject</span> <span>*</span><span>func_module</span><span>;</span>      <span>/* The __module__ attribute, can be anything */</span>
    <span>PyObject</span> <span>*</span><span>func_annotations</span><span>;</span> <span>/* Annotations, a dict or NULL */</span>
    <span>PyObject</span> <span>*</span><span>func_qualname</span><span>;</span>    <span>/* The qualified name */</span>
    <span>vectorcallfunc</span> <span>vectorcall</span><span>;</span>
<span>}</span> <span>PyFunctionObject</span><span>;</span>
</pre>
</div>
<h4>frame object</h4>
<p>When executing a code object, the VM has to keep track of the values of variables and the constantly changing value stack. It also needs to remember where it stopped executing the current code object to execute another and where to go on return. CPython stores this information inside a frame object, or simply a frame. A frame provides a state in which a code object can be executed. Since we're getting more accustomed with the source code, I leave the definition of the frame object here as well:</p>
<div>
<pre><span>struct</span> <span>_frame</span> <span>{</span>
    <span>PyObject_VAR_HEAD</span>
    <span>struct</span> <span>_frame</span> <span>*</span><span>f_back</span><span>;</span>      <span>/* previous frame, or NULL */</span>
    <span>PyCodeObject</span> <span>*</span><span>f_code</span><span>;</span>       <span>/* code segment */</span>
    <span>PyObject</span> <span>*</span><span>f_builtins</span><span>;</span>       <span>/* builtin symbol table (PyDictObject) */</span>
    <span>PyObject</span> <span>*</span><span>f_globals</span><span>;</span>        <span>/* global symbol table (PyDictObject) */</span>
    <span>PyObject</span> <span>*</span><span>f_locals</span><span>;</span>         <span>/* local symbol table (any mapping) */</span>
    <span>PyObject</span> <span>**</span><span>f_valuestack</span><span>;</span>    <span>/* points after the last local */</span>

    <span>PyObject</span> <span>**</span><span>f_stacktop</span><span>;</span>          <span>/* Next free slot in f_valuestack. ... */</span>
    <span>PyObject</span> <span>*</span><span>f_trace</span><span>;</span>          <span>/* Trace function */</span>
    <span>char</span> <span>f_trace_lines</span><span>;</span>         <span>/* Emit per-line trace events? */</span>
    <span>char</span> <span>f_trace_opcodes</span><span>;</span>       <span>/* Emit per-opcode trace events? */</span>

    <span>/* Borrowed reference to a generator, or NULL */</span>
    <span>PyObject</span> <span>*</span><span>f_gen</span><span>;</span>

    <span>int</span> <span>f_lasti</span><span>;</span>                <span>/* Last instruction if called */</span>
    <span>/* ... */</span>
    <span>int</span> <span>f_lineno</span><span>;</span>               <span>/* Current line number */</span>
    <span>int</span> <span>f_iblock</span><span>;</span>               <span>/* index in f_blockstack */</span>
    <span>char</span> <span>f_executing</span><span>;</span>           <span>/* whether the frame is still executing */</span>
    <span>PyTryBlock</span> <span>f_blockstack</span><span>[</span><span>CO_MAXBLOCKS</span><span>];</span> <span>/* for try and loop blocks */</span>
    <span>PyObject</span> <span>*</span><span>f_localsplus</span><span>[</span><span>1</span><span>];</span>  <span>/* locals+stack, dynamically sized */</span>
<span>};</span>
</pre>
</div>
<p>The first frame is created to execute a module's code object. CPython creates a new frame whenever it needs to execute another code object. Each frame has a reference to the previous frame. Thus, frames form a stack of frames, also known as the call stack, with the current frame sitting on top. When a function is called, a new frame is pushed on the stack. On return from the currently executing frame, CPython continues execution of the previous frame by remembering its last processed instruction. In some sense the CPython VM does nothing but constructs and executes the frames. But as we'll soon see, this summary, to put it mildly, hides some details.</p>
<h3>Threads, interpreters, runtime</h3>
<p>We've already looked at the three important concepts:</p>
<ul>
<li>a code object</li>
<li>a function object; and</li>
<li>a frame object.</li>
</ul>
<p>CPython has three more:</p>
<ul>
<li>a thread state</li>
<li>an interpreter state; and</li>
<li>a runtime state.</li>
</ul>
<h4>thread state</h4>
<p>A thread state is a data structure that contains thread-specific data including the call stack, an exception state and the debugging settings. It should not be confused with an OS thread. They're closely connected, though. Consider what happens when you use a standard <code>treading</code> module to run a function in a separate thread:</p>
<div>
<pre><span>from</span> <span>threading</span> <span>import</span> <span>Thread</span>

<span>def</span> <span>f</span><span>():</span>
    <span>"""Perform an I/O-bound task"""</span>
    <span>pass</span>

<span>t</span> <span>=</span> <span>Thread</span><span>(</span><span>target</span><span>=</span><span>f</span><span>)</span>
<span>t</span><span>.</span><span>start</span><span>()</span>
<span>t</span><span>.</span><span>join</span><span>()</span>
</pre>
</div>
<p><code>t.start()</code> actually creates a new OS thread by calling the OS function (<code>pthread_create</code> on UNIX-like systems and <code>_beginthreadex</code> on Windows). The newly created thread invokes the function from the <code>_thread</code> module that is responsible for calling the target. This function receives not only the target and the target's arguments but also a new thread state to be used within a new OS thread. An OS thread enters the evaluation loop with its own thread state, thus always having it at hand.</p>
<p>We may remember here the famous GIL (Global Interpreter Lock) that prevents multiple threads to be in the evaluation loop at the same time. The major reason for that is to protect the state of CPython from corruption without introducing more fine-grained locks. <a href="https://docs.python.org/3.9/c-api/index.html">The Python/C API Reference</a> explains the GIL clearly:</p>
<blockquote>
<p>The Python interpreter is not fully thread-safe. In order to support multi-threaded Python programs, there&rsquo;s a global lock, called the global interpreter lock or GIL, that must be held by the current thread before it can safely access Python objects. Without the lock, even the simplest operations could cause problems in a multi-threaded program: for example, when two threads simultaneously increment the reference count of the same object, the reference count could end up being incremented only once instead of twice.</p>
</blockquote>
<p>To manage multiple threads, there needs to be a higher-level data structure than a thread state.</p>
<h4>interpreter and runtime states</h4>
<p>In fact, there are two of them: an interpreter state and the runtime state. The need for both may not seem immediately obvious. However, an execution of any program has at least one instance of each and there are good reasons for that.</p>
<p>An interpreter state is a group of threads along with the data specific to this group. Threads share such things as loaded modules (<code>sys.modules</code>), builtins (<code>builtins.__dict__</code>) and the import system (<code>importlib</code>).</p>
<p>The runtime state is a global variable. It stores data that is specific to a process. This includes CPython state (is it initialized or not?) and the GIL mechanism.</p>
<p>Usually, all threads of a process belong to the same interpreter. There are, however, rare cases when one may want to create a subinterpreter to isolate a group of threads. <a href="https://modwsgi.readthedocs.io/en/develop/user-guides/processes-and-threading.html#python-sub-interpreters">mod_wsgi</a>, which uses distinct interpreters to run WSGI applications, is one example. The most obvious effect of isolation is that each group of threads gets its own version of all modules including <code>__main__</code>, which is a global namespace.</p>
<p>CPython doesn't provide an easy way to create new interpreters analogues to <code>threading</code> module. The feature is supported only via Python/C API, but <a href="https://www.python.org/dev/peps/pep-0554/">this may change</a> someday.</p>
<h3>Architecture summary</h3>
<p>Let's make a quick summary of the CPython's architecture to see how everything fits together. The interpreter can be viewed as a layered structure. The following sums up what the layers are:</p>
<ol>
<li>Runtime: global CPython state of a process; includes GIL and memory allocation mechanism</li>
<li>Interpreter: group of threads and some data they share such as imported modules.</li>
<li>Thread: data specific to a single OS thread; this includes call stack.</li>
<li>Frame: element of call stack; provides a state to execute a code object.</li>
<li>Evaluation loop: executes a code object, which tells what a code block does and contains bytecode and names of variables.</li>
</ol>
<p>The layers are represented by the corresponding data structures, which we've already seen. In some cases they are not equivalent, though. For example, the mechanism of memory allocation is implemented using global variables. It's not a part of the runtime state but certainly a part of CPython runtime layer.</p>
<h3>Conclusion</h3>
<p>In this part we've outlined what <code>python</code> does to execute a Python program. We've seen that it works in three stages:</p>
<ol>
<li>
<p>initializes Python runtime</p>
</li>
<li>
<p>compiles a source code to a module's code object; and</p>
</li>
<li>
<p>executes bytecode of the code object.</p>
</li>
</ol>
<p>The part of the interpreter that is responsible for bytecode execution is called a virtual machine. The CPython VM has several particularly important concepts: a code object, a frame object, a thread state, an interpreter state and the runtime. These data structures form a core of the CPython architecture.</p>
<p>We haven't covered a lot of things. We avoided digging into the source code. The initialization and compilation stages were completly out of our scope. Instead, we started with the broad overview of the VM. In this way, I think, we can better see the responsibilities of each stage. Now we know what CPython compiles source code to &ndash; to the code object. <a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-2-how-the-cpython-compiler-works/">Next time</a> we'll see how it does that.</p>
<p><br></p>
<p><em>If you have any questions, comments or suggestions, feel free to contact me at victor@tenthousandmeters.com</em></p>
<p><br></p>
<p><strong>Update 4 September 2020</strong>: I've made <a href="https://tenthousandmeters.com/materials/python-behind-the-scenes-a-list-of-resources/">a list of resources</a> that I've used to learn about CPython internals</p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via blog | TenThousandMeters.com<br>
<br>
December 14, 2020 at 02:49PM
